### DISCUSSION TITLE
Flitsmeister navigation SDK's merge to MapLibre

### DISCUSSION THREAD

**Discussion by SjoerdPerfors** [maplibre/maplibre#7] [Ideas]
Hi All,

I'm Sjoerd and Co-founder Flitsmeister. Flitsmeister is an app for on the go with a large user base (2.3mio MAU) in the Netherlands. These users are all using a complete license free Mapbox SDK for navigation and maps.

I want to start this discussion about transferring our Navigation SDK's to the GitHub account of MapLibre:

- https://github.com/flitsmeister/flitsmeister-navigation-ios
- https://github.com/flitsmeister/flitsmeister-navigation-android

Reason for this is to get a bigger group working on all the SDK's together.

**For iOS:**

1. we first need to integrate Maplibre GL
2. there is currently no navigation SDK available on Maplibre. So this could be a new repo.

**For Android:**

1. Also integrate Maplibre GL first
2. Find out which version exactly was forked by Maplibre (https://github.com/maplibre/maplibre-navigation-android) 
3. Take a diff between existing fork on Maplibre account to see what is changed
4. I think the most easiest thing is to do a PR to the existing repo, but to be defined.

**Timing:** 
We want to plan this work in Q1 but have some other projects running as well.

**Questions:**
What do you think of this plan/approach?
Who are the stakeholders to approve this plan/approach?
How to move forward?

Best regards,
Sjoerd

**Comment by petr-pokorny-1** [comment]
@SjoerdPerfors this is amazing. There is a MapLibre steering committee where these things are discussed on regular basis. I will ask guys there to invite you.

**Reply by SjoerdPerfors** [reply]
Received the invite, thanks.

**Comment by karussell** [comment]
This is great news!

> Find out which version exactly was forked by Maplibre (https://github.com/maplibre/maplibre-navigation-android)

Together with @boldtrn I'm the "owner" of this repo and I think this is not necessary and instead replace this with your code entirely and maybe some minor edits. There is also a fork that does the Maplibre integration, see this discussion: https://github.com/maplibre/maplibre-navigation-android/issues/8#issuecomment-907437506

So a nice start could be to delete everything in this repo and replace with your code (or the linked fork) and propose this as a PR :) (?)

**Reply by nyurik** [reply]
The last common revision seems to be [90451242](https://github.com/maplibre/maplibre-navigation-android/tree/90451242f884ce74cd54b9a51ddf416f075a24a5):  See [Maplibre diff](https://github.com/maplibre/maplibre-navigation-android/compare/90451242...master) and [flitsmeister diff](https://github.com/flitsmeister/flitsmeister-navigation-android/compare/90451242...master) (click "files changed" for details).

I do think it would be very strange to host two forks of the same code in the same organization -- I doubt we will have enough volunteers to maintain both in the long term, so I think these efforts should merge (afterall, the goal of transferring the code here is to gain more attention and contributors, not to split the community). I personally have no in-depth knowledge of either of the projects, so can't really say which path is better, or how (and if?) they should be merged, so I will leave that to the experts in the field.

**Reply by karussell** [reply]
Yes, probably my comment was ambiguous, but I proposed this, i.e. that the new code *replaces* the old via a PR, but uses the same repository. 

Maybe it would be cleaner, when we archive the current repo and give it a different name, so that the flitsmeister repo can get the "maplibre-navigation-android" name.

**Reply by boldtrn** [reply]
Yes I fully agree, replacing the current code with the flitsmeister code makes sense

**Comment by SjoerdPerfors** [comment]
We have updated our Navigation repo for Android to let it work with Maplibre maps sdk. Next is testing it in our app and have it released then we will be ready to move it's repo to Maplibre.

If anyone is interested: https://github.com/flitsmeister/flitsmeister-navigation-android/pull/14

**Reply by karussell** [reply]
Nice! Let me know when we should push the move to the Maplibre org forward. We could rename the maplibre-navigation-android repo to maplibre-navigation-android-old so that the name maplibre-navigation-android is free again :)

**Comment by Archdoog** [comment]
I'm starting to work on the same problem for our iOS and later Android mobile apps and would be interested in helping this effort on maplibre. I think there are likely to be more apps in the same boat between now and next April when Mapbox says they're sunsetting the old API/SDK to push apps onto the latest MAU based services.

Some of our goals are: 

1. Easy transition from android 0.42 & iOS 1.0.0 (this should be easy enough given they're not too far off the last fully open source sdks). 
2. With the advent of SwiftUI & Jetpack Compose, we're very interested in providing the building blocks to create a bespoke UI. On iOS we've already built our own in-house SwiftUI based navigation UI and have been playing around with just the core Router, etc. With this in mind, being able to easily select the navigation core, or the navigation core and navigation ui.
3. Choose your own mapping and routing. In our case we've already put a large effort into Mapbox's latest maps sdk and our own valhalla based routing service. By breaking down each navigation sdk into the correct modules, it seems reasonable to give apps the ability to incrementally adopt bit by bit of maplibre as they increment through API services whether for routing, tiles, etc.

Thoughts? 

I'm just about to start integrating the various pieces of navigation into our iOS app. It would work extremely well if my iOS work on top of mapbox's old iOS sdk was contributed directly to a maplibre-navigation-swift repo. My expertise is Swift/iOS, with a bit of Kotlin, JavaScript, C++ integration.

**Comment by ianthetechie** [comment]
I’d like to add my two cents here. Echoing what @Archdoog said above, I want to emphasize that this would serve everyone best if we can extract the core business logic of navigation from the rest and use a layered approach that lets users mix and match.

I come from the native mobile world and would naturally prefer to use something like SwiftUI + MapLibre GL Native, but I’m also seeing what looks like a big surge in interest in Flutter. Kotlin multi platform also shows some promise. And much as I’m not a fan, React Native isn’t likely to disappear anytime soon. Some enterprises use Xamarin as well. 

I don’t have a perfect solution in mind yet but I’d like to get a discussion going on what this core would look like. At a high level, it would need an API that allows any “native” location manager to send updates with the users location and heading, and then some mechanism for informing the caller of the current state. It would also need a common interface to the routing backend.

I have a few ideas about how all this would look at the API level, but it’s an open question in my mind how we write the common core. Given the diversity of platforms out there, I would suggest that it interface well with C, since that’s basically always a common denominator, and even RN can call out to native modules. Then build / generate platform specific bindings for easy integration.

I have not looked deeply into Kotlin Multiplattform, but if anyone has, do you know if it is easy to make a library that is callable easily from Swift on iOS, Kotlin on Android, Flutter with Dart bindings, etc.? I don’t want us to build something that only works with one framework.

We could build a safe, high quality library with a C interface using Rust, and then generate idiomatic bindings for other platforms. The downside is that I’m not sure the level of support for these platforms and how fast new CPUs are supported but if the library is minimal in nature I think we could be fine as rustc is LLVM based and we won’t need a lot of complex libraries. I can investigate this angle further if there is interest.

**Reply by ianthetechie** [reply]
This discussion is essentially closed now that it has actually been merged, but for anyone finding this later, I've collected my thoughts into a fuller proposal: https://stadiamaps.notion.site/Next-Gen-Navigation-SDK-f16f987bfa5a455296b0671636033cdb.

**Comment by Archdoog** [comment]
I'd be happy to help explore this on iOS w/ Swift and Android Kotlin. 

On iOS this should be relatively straightforward as long as everything's exposed through C and Swift can call it directly. I've worked on a similar project that created a Swift module for the full certificate generation process using OpenSSL's C library (CA, CSR, etc).
